pkg/segment/query/querystatus.go

```diff
@@ -995,3 +1017,92 @@ func SetQidAsFinishedForPipeRespQuery(qid uint64) {
 		rQuery.StateChan <- &QueryStateChanData{StateName: COMPLETE}
 	}
 }
+
+func InitProgressForRRCCmd(totalUnits uint64, qid uint64) {
+	arqMapLock.RLock()
+	rQuery, ok := allRunningQueries[qid]
+	arqMapLock.RUnlock()
+	if !ok {
+		log.Errorf("InitProgressForRRCCmd: qid %+v does not exist!", qid)
+		return
+	}
+
+	rQuery.rqsLock.Lock()
+	defer rQuery.rqsLock.Unlock()
+
+	if rQuery.isAsync {
+		if rQuery.Progress != nil {
+			return
+		}
+
+		rQuery.Progress = &structs.Progress{
+			TotalUnits:   totalUnits,
+			TotalRecords: rQuery.totalRecsToBeSearched,
+		}
+	}
+}
+
+func IncProgressForRRCCmd(recordsSearched uint64, unitsSearched uint64, qid uint64) error {
+	arqMapLock.RLock()
+	rQuery, ok := allRunningQueries[qid]
+	arqMapLock.RUnlock()
+	if !ok {
+		return putils.TeeErrorf("IncProgressForRRCCmd: qid %+v does not exist!", qid)
+	}
+
+	rQuery.rqsLock.Lock()
+	defer rQuery.rqsLock.Unlock()
+
+	if rQuery.isAsync {
+		if rQuery.Progress == nil {
+			return putils.TeeErrorf("IncProgressForRRCCmd: Progress is not initialized! for qid=%v", qid)
+		}
+
+		rQuery.Progress.UnitsSearched += unitsSearched
+		rQuery.Progress.RecordsSearched += recordsSearched
+
+		wsResponse := CreateWSUpdateResponseWithProgress(qid, rQuery.QType, rQuery.Progress)
+		rQuery.StateChan <- &QueryStateChanData{
+			StateName:    QUERY_UPDATE,
+			UpdateWSResp: wsResponse,
+		}
+	}
+
+	return nil
+}
+
+func GetProgress(qid uint64) (structs.Progress, error) {
+	arqMapLock.RLock()
+	rQuery, ok := allRunningQueries[qid]
+	arqMapLock.RUnlock()
+	if !ok {
+		return structs.Progress{}, putils.TeeErrorf("GetProgress: qid %+v does not exist!", qid)
+	}
+
+	rQuery.rqsLock.Lock()
+	defer rQuery.rqsLock.Unlock()
+	if rQuery.Progress == nil {
+		return structs.Progress{}, putils.TeeErrorf("GetProgress: Progress is not initialized! for qid=%v", qid)
+	}
+
+	return structs.Progress{
+		TotalUnits:      rQuery.Progress.TotalUnits,
+		UnitsSearched:   rQuery.Progress.UnitsSearched,
+		TotalRecords:    rQuery.Progress.TotalRecords,
+		RecordsSearched: rQuery.Progress.RecordsSearched,
+	}, nil
+}
+
+func CreateWSUpdateResponseWithProgress(qid uint64, qType structs.QueryType, progress *structs.Progress) *structs.PipeSearchWSUpdateResponse {
+	percComplete := float64(0)
+	if progress.TotalUnits > 0 {
+		percComplete = (float64(progress.UnitsSearched) * 100) / float64(progress.TotalUnits)
+	}
```

Comment:
I think this is going to cause the percent complete value to be inaccurate when the query exits a long time before searching all blocks/segments. For example, after time-filtering and CMI, we may have 100 segments to search, but we stop the query once we get enough results (100 on the UI), and that could happen in the first 5 segments. Then the percent complete would be <=5% and then jump to 100%; it would be better if e.g., when we returned the first 30 results to the UI, we mark the query as 30% complete. Or maybe we need to track both the percent complete we are on searching segments and the percent complete on returning enough results to the UI; since we'll end the query once one of those reaches 100%, we take the max of them and use that as the percent complete

Reply:
addressed in the latest commit.